### ISSUE TITLE
Improved API for workflows

### ISSUE THREAD

**Issue by oyvindeide** [equinor/ert#1804]
Users currently making workflows for `ert` are given the full `EnkfMain` object to manipulate, both for fetching data and posting data. This is not really in line with the current development, as the workflow should be oblivious to where the data is coming from, and we should have a good API for interacting with the data. It would also be good to have different types of workflows and not a one solution fits all.

We should consider moving some of the complexity from the workflow author into `ert`, and in that way we can more easily swap where the data is coming from, utilizing the new storage solution as soon as it is available. That way we also get user testing as soon as possible for a larger range of cases, and we start moving in the direction we would like to go.

A good starting point could be to expose some of the functionality from https://github.com/equinor/ert/blob/a4d473fc6caf8af560b3751387537a7c520b8111/ert_shared/libres_facade.py#L14 and let `ert` decide where the data should be served from. As we are running the function we could for example do something similar to `fixtures` in `pytest`. This will include a small refactor of existing functionality, but should be doable:

```python
import abc
import functools
import inspect


def do_crazy_ert_thing():
    return 42

FIXTURES = {
    "observation_keys": ["obs_1", "obs_2"],
    "parameter_names": ["a", "b"],
    "some_easy_yet_complicated_operation": do_crazy_ert_thing()
}


def inject_fixtures(func):
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        extra_args = inspect.signature(self.run).parameters.keys()
        for arg in extra_args:
            if arg in FIXTURES:
                kwargs[arg] = FIXTURES[arg]
        return func(self, *args, **kwargs)
    return wrapper


class ErtScript(abc.ABC):
    def __init__(self, ert):
        self._ert = ert

    @abc.abstractmethod
    def run(self, *args, **kwargs):
        pass

    @inject_fixtures
    def init_and_run(self, *args, **kwargs):
        self.run(*args, **kwargs)

    def ert(self):
        return self._ert

class MyClass(ErtScript):
    def run(self, observation_keys):
        print(f"obs: {observation_keys}")


for script in AbstractScript.__subclasses__():
    script("EnkFMain").init_and_run()
```

This also has the benefit of abstracting away the different types of observations, parameters, etc, and pulling us in the direction we would like to go. It will also allow us to work on the API iteratively.

At the same time we could also mature and test concepts that will work in `ert3` (for example different types of workflows).

TL;DR:
* Abstract away data fetching
* Improved API
* Mature concept of workflows for `ert3`

This is a short(ish) summary of a discussion, anything to add @markusdregi @oysteoh @ManInFez or @xjules?

**Comment by jondequinor** [comment]
Also see https://github.com/equinor/ert/issues/719

Is it possible to do this from a clean slate? Put a proxy in front of the workflow system and implement these proposed changes as a completely new workflow system (named something else, of course).

**Comment by oyvindeide** [comment]
I think a clean slate implementation could be done side by side with this, but for a workflow system to be viable we probably need the existing one until the new storage solution is matured a bit more. I think a limited iteration on the existing solution could give us valuable insight into what to bring forward in a new solution.

But something to consider for sure!
